File: app/src/main/java/com/by5388/sy95306v2/main/presenter/MainPresenter.java
package com.by5388.sy95306v2.main.presenter;

import android.util.Log;

import com.by5388.sy95306v2.database.IShenYangDbApi;
import com.by5388.sy95306v2.exception.NetworkException;
import com.by5388.sy95306v2.main.model.IMainModel;
import com.by5388.sy95306v2.main.model.MainModel;
import com.by5388.sy95306v2.main.view.IMainView;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

/**
 * @author by5388  on 2019/1/2.
 */
public class MainPresenter implements IMainPresenter {
    private static final String TAG = MainPresenter.class.getSimpleName();

    private final IMainView mMainView;
    private final IMainModel mModel;
    private final CompositeDisposable mDisposable;

    public MainPresenter(IMainView view) {
        this.mMainView = view;
        this.mModel = new MainModel();
        this.mDisposable = new CompositeDisposable();
    }

    @Override
    public void checkUpdate() {
        mMainView.onStartChecking();
        mDisposable.add(mModel.isNeedUpdate()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(isNeedUpdate -> {
                    if (isNeedUpdate) {
                        mMainView.onNotifyUpdate();
                    } else {
                        mMainView.onFinishChecked();
                    }
                }, throwable -> {
                    mMainView.openNetWorkSetting();
                    mMainView.onTip("请重试");
                })
        );

    }

    @Override
    public void startUpdate() {
        mMainView.onShowUpdating();
        // TODO: 2019/11/26 对异常没有处理，导致闪退
        mDisposable.add(mModel.getStationCount()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(mInteger -> {
                    mMainView.updateAllCount(mInteger);
                    clearData();
                }, throwable -> {
                    if (throwable instanceof NetworkException) {
                        Log.e(TAG, "startUpdate: ", throwable);
                        mMainView.onTip(throwable.getLocalizedMessage());
                    }
                })
        );

    }

    private void clearData() {
        mDisposable.add(mModel.clearData()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(result -> {
                    if (IShenYangDbApi.ACTION_DELETE_SUCCESS == result) {
                        insertData();
                    }
                })
        );
    }

    private void insertData() {
        mDisposable.add(mModel.insertData()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(mMainView::updateProgress,
                        throwable -> mMainView.onTip("异常"),
                        () -> {
                            mMainView.onFinishUpdate();
                            mModel.onFinishUpdate();
                        })
        );
    }
}

File: sy95306v1kotlin/src/main/java/com/by5388/sy95306/kotlin/main/presenter/MainPresenter.kt
package com.by5388.sy95306.kotlin.main.presenter

import android.util.Log
import com.by5388.sy95306.kotlin.R
import com.by5388.sy95306.kotlin.main.model.IMainModel
import com.by5388.sy95306.kotlin.main.model.MainModel
import com.by5388.sy95306.kotlin.main.view.IMainView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

/**
 * @author  by5388  on 2019/1/10.
 */
class MainPresenter(private val view: IMainView) : IMainPresenter {
    private val model: IMainModel
    private val disposable: CompositeDisposable

    init {
        model = MainModel()
        disposable = CompositeDisposable()
    }

    companion object {
        private const val TAG = "MainPresenter"
    }

    override fun checkUpdate() {
        // TODO: 操作前 检查网络状态
        disposable.add(model.checkUpdate()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (it) {
                        view.notifyUpdate()
                    } else {
                        view.tip(R.string.is_update_data)
                    }
                }, {
                    view.tip(R.string.error_try_again)
                }))
    }

    override fun startUpdate() {
        disposable.add(model.getStationCount()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    view.showUpdateDiaglog(it)
                    clearData()
                }, { view.tip(R.string.error_try_again) })
        )
    }

    override fun unSubject() {
        disposable.clear()
    }

    private fun clearData() {
        val errorCode = -1
        disposable.add(model.clearData()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (errorCode != it) {
                        insertData()
                    }
                }, {
                    view.tip(R.string.error_try_again)
                })
        )
    }

    private fun insertData() {
        disposable.add(model.insertStationList()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    Log.d(TAG, "insertData = $it")
                    view.updateProgress(it)
                }, {
                    view.tip(R.string.error_try_again)
                }, {
                    Log.d(TAG, "加载完成")
                    view.finishUpdate()
                    model.finishUpdate()
                })
        )
    }
}